### ISSUE TITLE
SHAP algorithm

### ISSUE THREAD

**Issue by easyans** [easyans/Explainable-AI-for-Early-Detection-and-Severity-Assessment-of-Alzheimer-s-Disease#1]
`# %% Imports
from utils import DataLoader
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score, accuracy_score
import shap

# %% Load and preprocess data
data_loader = DataLoader()
data_loader.load_dataset()
data_loader.preprocess_data()
# Split the data for evaluation
X_train, X_test, y_train, y_test = data_loader.get_data_split()
# Oversample the train data
X_train, y_train = data_loader.oversample(X_train, y_train)
print(X_train.shape)
print(X_test.shape)

# %% Fit blackbox model
rf = RandomForestClassifier()
rf.fit(X_train, y_train)
y_pred = rf.predict(X_test)
print(f"F1 Score {f1_score(y_test, y_pred, average='macro')}")
print(f"Accuracy {accuracy_score(y_test, y_pred)}")

# %% Create SHAP explainer
explainer = shap.TreeExplainer(rf)
# Calculate shapley values for test data
start_index = 1
end_index = 2
shap_values = explainer.shap_values(X_test[start_index:end_index])
X_test[start_index:end_index]

# %% Investigating the values (classification problem)
# class 0 = contribution to class 1
# class 1 = contribution to class 2
print(shap_values[0].shape)
shap_values

# %% >> Visualize local predictions
shap.initjs()
# Force plot
prediction = rf.predict(X_test[start_index:end_index])[0]
print(f"The RF predicted: {prediction}")
shap.force_plot(explainer.expected_value[1],
                shap_values[1],
                X_test[start_index:end_index], # for values
)

# %% >> Visualize global features
# Feature summary
shap.summary_plot(shap_values, X_test)
# %%

![Image](https://github.com/easyans/Explainable-AI-for-Early-Detection-and-Severity-Assessment-of-Alzheimer-s-Disease/assets/82706959/c85f3c79-425f-4508-aaf9-e85cc5ee9d72)
![Image](https://github.com/easyans/Explainable-AI-for-Early-Detection-and-Severity-Assessment-of-Alzheimer-s-Disease/assets/82706959/642e7cfb-b470-4f9d-b3ee-1e9658bfc1e5)

![Image](https://github.com/easyans/Explainable-AI-for-Early-Detection-and-Severity-Assessment-of-Alzheimer-s-Disease/assets/82706959/f6dd2a57-2a66-44f4-8c63-50a7130a86f9)

`
